service/src/main/java/org/dice/FactCheck/Corraborative/NPMICalculator.java

```diff
@@ -13,331 +12,558 @@
-        double npmi;
-
-        // Path and predicate never occur together
-        if (count_path_Predicate_Occurrence == 0) {
-            // Since we know that A and B exist, there is a chance that they should occur
-            // together. Since it never happens, we have to return -1
-            //npmi = -1;
-            npmi = 0;
-        } else {
-            npmi = calculateNPMI(Math.log(count_path_Predicate_Occurrence), logSubObjTriples,
-                    Math.log(count_Path_Occurrence), logSubObjTriples, Math.log(count_predicate_Occurrence),
-                    logSubObjTriples);
-        }
-        if ((filter != null) && (!filter.npmiIsOk(npmi, pathLength, count_path_Predicate_Occurrence,
-                count_Path_Occurrence, count_predicate_Occurrence))) {
-            throw new NPMIFilterException("The NPMI filter rejected the calculated NPMI.");
-        }
-        return npmi;
+    if (pathLength == 2) {
+      String firstPath = generatePath(querySequence, 0);
+      String secondPath = generatePath(querySequence, 1);
+
+      return "Select (sum(?b1*?b2) as ?c) where {\n"
+          + "select (count(*) as ?b2) ?b1 where { \n"
+          + firstPath
+          + " .\n"
+          + subTypeTriples
+          + "{ \n"
+          + "select (count(*) as ?b1) ?x1 where { \n"
+          + secondPath
+          + " .\n"
+          + objTypeTriples
+          + "} group by ?x1\n"
+          + "}\n"
+          + "} group by ?b1\n"
+          + "}\n";
     }
 
-    /**
-     * Calculates the NPMI value for an event A, an event B and their common
-     * occurrence AB based on the given <b>logarithmic</b> count values.
-     * 
-     * @param logCountAB the number of common occurrences of A and B.
-     * @param logNormAB  the theoretical maximum of the common occurrence of A and
-     *                   B.
-     * @param logCountA  the occurrence count of A.
-     * @param logNormA   the theoretical maximum of the occurrence count of A.
-     * @param logCountB  the occurrence count of B.
-     * @param logNormB   the theoretical maximum of the occurrence count of B.
-     * @return the NPMI value
-     */
-    public static double calculateNPMI(double logCountAB, double logNormAB, double logCountA, double logNormA,
-            double logCountB, double logNormB) {
-        // Calculate probabilities
-        double logProbA = logCountA - logNormA;
-        double logProbB = logCountB - logNormB;
-        double logProbAB = logCountAB - logNormAB;
-
-        // If the probability of AB is 1.0 (i.e., its log is 0.0)
-        if (logProbAB == 0) {
-            return 1.0;
-        } else {
-            return (logProbAB - logProbA - logProbB) / -logProbAB;
-        }
+    if (pathLength == 3) {
+      String firstPath = generatePath(querySequence, 0);
+      String secondPath = generatePath(querySequence, 1);
+      String thirdPath = generatePath(querySequence, 2);
+
+      return "select (sum(?b3*?k) as ?c) where { \n"
```

Comment:
Just as a hint: working with a StringBuilder is much more efficient than the simple `+` operator :wink: 

So this piece of code should be something like:
```java
return new StringBuilder()
.append("select (sum(?b3*?k) as ?c) where { \n"
             + "select (count(*) as ?b3) (?b2*?b1 as ?k) ?x1 where { \n")
.append(firstPath)
.append(" .\n")
.append(subTypeTriples)
.append("{ \n"
             + "Select (count(*) as ?b2) ?x1 ?b1 where { \n")
.append(secondPath)
.append("{ \n"
             + "select (count(*) as ?b1) ?x2 where { \n")
.append(thirdPath)
.append(". \n")
.append(objTypeTriples)
.append("} group by ?x2\n"
             + "}\n"
             + "} group by ?b1 ?x1\n"
             + "}\n"
             + "} group by ?x1 ?b2 ?b1\n"
             + "}\n").toString();
```
The plus operations that I left in there should be already executed by the Java compiler, if I remember correctly :thinking:

Reply:
I think in this case the string builder doesn't have meaning full performance enhancement but I will check.

Reply:
Mhh... it seems like the compiler is able to do this work for us. [This discussion](https://stackoverflow.com/questions/1532461/stringbuilder-vs-string-concatenation-in-tostring-in-java) refers to the statement in [this description](https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.18.1)  (the one with the tiny font). It mainly states that the compiler can replace simple String concatenation by using a StringBuffer. However, using StringBuilder is in our case even better than using the StringBuffer :thinking: 

However, it is just something that we may want to take into consideration in the future. It is not necessary to change it for this PR. Maybe some student assistant task :smile:

Reply:
I saw your commit but the comparison does not seem to be fair. We need to take into account that the `+` operator is already executed by the compiler in case the two Strings are constants. This is the case in our example and would bias the comparison towards preferring the `+` operator.

I checked it using the following simple main method which takes care of this issue (i.e., use `+` for constants and the `append` method for combining variables with constants):
```java
    public static void main(String[] args) {
        int runs = 1000;
        String values[] = new String[runs * 5];
        // Generate some dynamic values
        for (int i = 0; i < values.length; ++i) {
            values[i] = "value" + i;
        }
        int valueId = 0;
        long pTimes[] = new long[10];
        long bTimes[] = new long[10];
        for (int i = 0; i < 10; ++i) {
            valueId = 0;
            pTimes[i] = System.nanoTime();
            for (int j = 0; j < runs; ++j) {
                String s1 =
                        "select (sum(?b3*?k) as ?c) where { \n"
                        + "select (count(*) as ?b3) (?b2*?b1 as ?k) ?x1 where { \n"
                        + values[valueId++]
                        + " .\n"
                        + values[valueId++]
                        + "{ \n"
                        + "Select (count(*) as ?b2) ?x1 ?b1 where { \n"
                        + values[valueId++]
                        + "{ \n"
                        + "select (count(*) as ?b1) ?x2 where { \n"
                        + values[valueId++]
                        + ". \n"
                        + values[valueId++]
                        + "} group by ?x2\n"
                        + "}\n"
                        + "} group by ?b1 ?x1\n"
                        + "}\n"
                        + "} group by ?x1 ?b2 ?b1\n"
                        + "}\n";
            }
            pTimes[i] = System.nanoTime() - pTimes[i];
            // StringBuilder
            valueId = 0;
            bTimes[i] = System.nanoTime();
            for (int j = 0; j < 100; ++j) {
                String s2 = new StringBuilder().append(
                        "select (sum(?b3*?k) as ?c) where { \n" +
                        "select (count(*) as ?b3) (?b2*?b1 as ?k) ?x1 where { \n").append(
                        values[valueId++]).append(
                        " .\n").append(
                        values[valueId++]).append(
                        "{ \n" +
                        "Select (count(*) as ?b2) ?x1 ?b1 where { \n").append(
                        values[valueId++]).append(
                        "{ \n" +
                        "select (count(*) as ?b1) ?x2 where { \n").append(
                        values[valueId++]).append(
                        ". \n").append(
                        values[valueId++]).append(
                        "} group by ?x2\n" +
                        "}\n" +
                        "} group by ?b1 ?x1\n" +
                        "}\n" +
                        "} group by ?x1 ?b2 ?b1\n" +
                        "}\n").toString();
            }
            bTimes[i] = System.nanoTime() - bTimes[i];
            // Ask the JVM to cleanup
            System.gc();
        }
        long avg = 0;
        System.out.println("With +");
        for (int j = 0; j < pTimes.length; ++j) {
            System.out.printf("%10d%n", pTimes[j]);
            avg += pTimes[j];
        }
        System.out.println("----------");
        System.out.printf("%8.1f%n", avg / (double) pTimes.length);
        System.out.println();
        avg = 0;
        System.out.println("With sb");
        for (int j = 0; j < bTimes.length; ++j) {
            System.out.printf("%10d%n", bTimes[j]);
            avg += bTimes[j];
        }
        System.out.println("----------");
        System.out.printf("%8.1f%n", avg / (double) bTimes.length);
    }
```
The output is as follows:
```
With +
   3601981
   2237661
   1501217
    681779
    714666
    931667
    634376
    593842
    595387
    653168
----------
 1214574.4

With sb
    321045
    227733
    187645
     48710
     50032
     71565
     44994
     45384
     44410
    106878
----------
  114839.6
```
As soon as you have a lot of these operations, the `StringBuilder` is faster :wink:

Reply:
Yes, I see, thank you for this test, I agree with you the StringBuilder is faster. In a loop situation, it would be a lot faster, I didn't consider that, this loop can be a large loop 
```
for (Future<Result> result : executor.invokeAll(pmiCallables)) 
```
and I considered it always should be under 20 iterations.